### PR TITLE
refactor(changelog): better typing, list comprehension

### PR DESCRIPTION
**Pull request by bearomorphism**
## Checklist

- [ ] I have read the [contributing guidelines](https://commitizen-tools.github.io/commitizen/contributing/)

### Code Changes

- [ ] Add test cases to all the changes you introduce
- [ ] Run `poetry all` locally to ensure this change passes linter check and tests
- [ ] Manually test the changes:
  - [ ] Verify the feature/bug fix works as expected in real-world scenarios
  - [ ] Test edge cases and error conditions
  - [ ] Ensure backward compatibility is maintained
  - [ ] Document any manual testing steps performed
- [ ] Update the documentation for the changes

### Documentation Changes

- [ ] Run `poetry doc` locally to ensure the documentation pages renders correctly

### CODE REVIEW

File: tests/test_changelog.py [hunk 1]
```diff
@@ -1219,22 +1219,22 @@ def test_order_changelog_tree(change_type_order, expected_reordering):
...
         else:
-            assert [*entry["changes"].keys()] == [*tree[index]["changes"].keys()]
+            assert [*entry["changes"].keys()] == [*entry["changes"].keys()]
 
 
 def test_order_changelog_tree_raises():
     change_type_order = ["BREAKING CHANGE", "feat", "refactor", "feat"]
     with pytest.raises(InvalidConfigurationError) as excinfo:
-        changelog.order_changelog_tree(COMMITS_TREE, change_type_order)
+        list(changelog.order_changelog_tree(COMMITS_TREE, change_type_order))
```
Comment (bearomorphism) [inline tests/test_changelog.py:1237]: Added `list` here for triggering the error.

File: commitizen/changelog.py [hunk 2]
```diff
@@ -187,24 +187,27 @@ def process_commit_message(
             changes[change_type].append(msg)
 
 
-def order_changelog_tree(tree: Iterable, change_type_order: list[str]) -> Iterable:
+def order_changelog_tree(
+    tree: Iterable[Mapping[str, Any]], change_type_order: list[str]
+) -> Generator[dict[str, Any], None, None]:
```
Comment (Lee-W) [inline commitizen/changelog.py:192]: If we want to make this type change, we probably would like to apply it on `generate_tree_from_commits`?

Comment (bearomorphism) [reply commitizen/changelog.py:192]: fixed

File: commitizen/changelog.py [hunk 3]
```diff
@@ -187,24 +187,24 @@ def process_commit_message(
...
-            if ct in entry["changes"]
-        ]
-        sorted_tree.append({**entry, **{"changes": OrderedDict(changes)}})
-    return sorted_tree
+        yield {
+            **entry,
+            "changes": OrderedDict(
+                (ct, entry["changes"][ct])
+                for ct in change_type_order
+                + sorted(set(entry["changes"].keys()) - set(change_type_order))
```
[the remark this answers is not in the document]
Comment (bearomorphism) [reply commitizen/changelog.py]: fixed.

**Review by Lee-W — APPROVED** [review]

**Comment by Lee-W** [comment]
@woile @noirbizarre I'm happy about this change. I'm planning on merging this one this Wednesday, but would appreciate a second eye on this one 🙂

**Comment by Lee-W** [comment]
I'll go ahead an merge it 💪
